# Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in projects 1 and 2.

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import pylab as pl
import numpy as np


In [2]:
df_raw = pd.read_csv("../assets/admissions.csv")
df = df_raw.dropna() 
print df.head()

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### 1. Let's create a frequency table of our variables

In [29]:
df.count()# frequency table for prestige and whether or not someone was admitted


admit           397
gre             397
gpa             397
prestige        397
prestige_1.0    397
prestige_2.0    397
prestige_3.0    397
prestige_4.0    397
dtype: int64

In [28]:
df.count()['admit'] == 1

False

## Part 2. Return of dummy variables

#### 2.1 Create class or dummy variables for prestige 

In [4]:
prestige_dummies = pd.get_dummies(df.prestige, prefix = "prestige")
df = pd.concat([df, prestige_dummies], axis = 1)

admit    gre   gpa  prestige  prestige_1.0  prestige_2.0  prestige_3.0  \
0      0  380.0  3.61       3.0             0             0             1   
1      1  660.0  3.67       3.0             0             0             1   

   prestige_4.0  
0             0  
1             0

In [5]:
df.head(10)

admit    gre   gpa  prestige  prestige_1.0  prestige_2.0  prestige_3.0  \
0      0  380.0  3.61       3.0             0             0             1   
1      1  660.0  3.67       3.0             0             0             1   
2      1  800.0  4.00       1.0             1             0             0   
3      1  640.0  3.19       4.0             0             0             0   
4      0  520.0  2.93       4.0             0             0             0   
5      1  760.0  3.00       2.0             0             1             0   
6      1  560.0  2.98       1.0             1             0             0   
7      0  400.0  3.08       2.0             0             1             0   
8      1  540.0  3.39       3.0             0             0             1   
9      0  700.0  3.92       2.0             0             1             0   

   prestige_4.0  
0             0  
1             0  
2             0  
3             1  
4             1  
5             0  
6             0  
7             0  
8             0  
9             0

#### 2.2 When modeling our class variables, how many do we need? 



Answer: 1

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

In [7]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(prestige_dummies.ix[:, 'prestige_1':])
print handCalc.head()

   admit    gre   gpa  prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             0             1             0
1      1  660.0  3.67             0             0             1             0
2      1  800.0  4.00             1             0             0             0
3      1  640.0  3.19             0             0             0             1
4      0  520.0  2.93             0             0             0             1


In [32]:
pd.crosstab(df['prestige_1.0'], df['admit'], margins=True)
# frequency table cutting prestige and whether or not someone was admitted


admit           0    1  All
prestige_1.0               
0             243   93  336
1              28   33   61
All           271  126  397

#### 3.1 Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college

In [61]:
rank_1 = 33/float(61)
print rank_1


0.540983606557


#### 3.2 Now calculate the odds of admission if you did not attend a #1 ranked college

In [60]:
rank_0 = 93/float(336)
print rank_0

0.276785714286


#### 3.3 Calculate the odds ratio

In [62]:
print rank_1/rank_0

1.95452141724


#### 3.4 Write this finding in a sentenance: 

Answer: The probability of a student getting admitted if they have attended a #1 ranked college is 0.54 while the probability of a student getting admitted if they did not attend a #1 ranked college is 0.28. This shows that it is tougher to get admitted if you have not attended a #1 ranked college. 

#### 3.5 Print the cross tab for prestige_4

In [63]:
pd.crosstab(df['prestige_4.0'], df['admit'], margins=True)

admit           0    1  All
prestige_4.0               
0             216  114  330
1              55   12   67
All           271  126  397

#### 3.6 Calculate the OR 

In [64]:
og = 12/float(67)
print og

0.179104477612


In [66]:
to = 114/float(330)
print to

0.345454545455


In [68]:
odd_ratio = og/to
print odd_ratio

0.518460329929


#### 3.7 Write this finding in a sentence

Answer: The probability of getting admitted when you have not attended a prestige 4.0 school is higher compared to if you have attended a prestige 4.0 college as shown by the figures of 0.345 and 0.179.

## Part 4. Analysis

In [70]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(prestige_dummies.ix[:, 'prestige_2':])
print data.head()

   admit    gre   gpa  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             1             0
1      1  660.0  3.67             0             1             0
2      1  800.0  4.00             0             0             0
3      1  640.0  3.19             0             0             1
4      0  520.0  2.93             0             0             1


We're going to add a constant term for our Logistic Regression. The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [71]:
# manually add the intercept
data['intercept'] = 1.0

#### 4.1 Set the covariates to a variable called train_cols

In [73]:
train_cols = ['gre','gpa','prestige_2.0','prestige_3.0', 'prestige_4.0']

#### 4.2 Fit the model

In [78]:
logit = sm.Logit(data['admit'], data[train_cols])
results = logit.fit()

Optimization terminated successfully.
         Current function value: 0.589121
         Iterations 5


#### 4.3 Print the summary results

In [79]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      392
Method:                           MLE   Df Model:                            4
Date:                Thu, 29 Jun 2017   Pseudo R-squ.:                 0.05722
Time:                        18:19:05   Log-Likelihood:                -233.88
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                 1.039e-05
================================================================================
                   coef    std err          z      P>|z|      [95.0% Conf. Int.]
--------------------------------------------------------------------------------
gre              0.0014      0.001      1.308      0.191        -0.001     0.003
gpa             -0.1323      0.195     -0.680      0.497        -0.514     0.249
prestige_2.0    -0.9562      0.302     -3.171      0.002        -1.547    -0.365
prestige_3.0    -1.5375      0.332     -4.627      0.000        -2.189    -0.886
prestige_4.0    -1.8699      0.401     -4.658      0.000        -2.657    -1.083
================================================================================
"""

#### 4.4 Calculate the odds ratios of the coeffiencents and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params
        
           conf.columns = ['2.5%', '97.5%', 'OR']

In [85]:
np.exp(results.params)

gre             1.001368
gpa             0.876073
prestige_2.0    0.384342
prestige_3.0    0.214918
prestige_4.0    0.154135
dtype: float64

In [92]:
params = results.params
conf = results.conf_int()
conf['OR'] = params
conf.columns = ['95.0%' , '25.0', '50.0']
print np.exp(conf)

                 95.0%      25.0      50.0
gre           0.999320  1.003420  1.001368
gpa           0.598303  1.282800  0.876073
prestige_2.0  0.212826  0.694082  0.384342
prestige_3.0  0.112055  0.412207  0.214918
prestige_4.0  0.070176  0.338540  0.154135


#### 4.5 Interpret the OR of Prestige_2

Answer: We can say that the odds of being admitted will decrease by about 0.38% if the prestige of a school the student attended is 2.

#### 4.6 Interpret the OR of GPA

Answer: We can say that the odds of being admitted depends heavily on a student's GPA. 

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values. This will allow us to see how the predicted probability of admission increases/decreases across different variables. First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa". This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [99]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [100]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max 
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)
print gres
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])
gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)
print gpas
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])


# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))

[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]
[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


#### 5.1 Recreate the dummy variables

In [103]:
prestige_dummies_2 = pd.get_dummies(combos["prestige"], prefix = "prestige")



# recreate the dummy variables

# keep only what we need for making predictions


KeyError: 'prestige'

#### 5.2 Make predictions on the enumerated dataset

#### 5.3 Interpret findings for the last 4 observations

Answer: 

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.